### PR TITLE
Make auto-replacement optional

### PR DESCRIPTION
**Pull request by sampsyo**
This is an alternative to #10 that makes things a little simpler. Now, there is still only one module file, but you can supply `?auto=off` to disable the auto-replacement feature.

What do you think about this alternative, @octogonz? Would this work for your SPA use case?

### CODE REVIEW

**Comment by octogonz** [comment]
No, this doesn't seem like it helps.

Our websites are built using [Docusaurus](https://docusaurus.io/), probably the most popular documentation system for open source projects.  Docusaurus is pretty complex web application that uses Webpack server-side rendering to create a separate .js file for each documentation page, which are dynamically loaded from a static CDN (without requiring a Node.js server).

What we need is a JavaScript module that can be bundled into a Webpack chunk and wrapped in a React component that will render the DOM nodes during React rendering.

To be clear:
- We really do not want to load the `emfed` bundle from someone else's CDN domain
- We'd prefer not to deploy `emfed`  with its own dedicated URL
- In other words, we want a conventional API function that Webpack can bundle, not a magical runtime analysis of `import.meta.url`

TBH maybe it would be better for me to do the work to reimplement the `emfed` concept as a proper React component, and then we would also get professional CSS styling instead of global CSS templates.  I just was lazy and hoping to avoid owning the maintenance of Mastodon protocol changes and security scrubbing. 😄

File: emfed.ts [hunk 1]
```diff
@@ -184,5 +184,22 @@ async function loadToots(element: Element) {
...
+}
+
+// By default, we automatically transform all links. But users can disable
+// this by supplying ?auto=off and instead manually invoking the API to
+// transform specific elements.
+if (
+  !(
+    import.meta &&
+    import.meta.url &&
+    new URL(import.meta.url).searchParams.get("auto") === "off"
```
Comment (octogonz) [inline emfed.ts:201]: Webpack's compile time tree-shaker is not going to be able to evaluate this condition.

Comment (sampsyo) [reply emfed.ts:201]: Forgive me if I'm misunderstanding something, but my original thought here was that a bundler would not need to evaluate the condition—it would just need to preserve it, so that it evaluates to `false` at run time. Does that make any sense?

Comment (sampsyo) [reply emfed.ts:201]: Oh, that was dumb; I of course was getting the condition backward in my head. We would want this to "default to off" instead of "defaulting to on." Sorry about that.

**Comment by sampsyo** [comment]
Cool; all that makes sense. If this kind of `import.meta` trickery is unworkable for bundlers, your original idea to split the two modules is probably the way to go.

While I definitely support a fork (I am obviously not a real frontend developer and I'm sure other people could do better), I'm also interested in iteratively improving this. So I'll probably forge ahead with your module-splitting idea anyway, even if you end up going in a different direction. (If you do end up building something separate, please feel free to steal anything you like from this project!)

File: emfed.ts (hunk 1 above)
Comment (octogonz) [reply emfed.ts:201]: Even if the condition defaults to false, the `new URL(import.meta.url)` code would still need to get executed at runtime (if I understand correctly) which seems janky.

**Comment by octogonz** [comment]
> TBH maybe it would be better for me to do the work to reimplement the `emfed` concept as a proper React component, and then we would also get professional CSS styling instead of global CSS templates. I just was lazy and hoping to avoid owning the maintenance of Mastodon protocol changes and security scrubbing.

I went ahead and did this, and it turned out to be a pretty useful exercise. 

Here's the result (with screenshots): https://github.com/microsoft/rushstack-websites/pull/204

Given that my case is already using React, the HTML generation simplified greatly. But relying on React probably works against your goal of providing a very tiny bundle.

I also started to think that it's important that the embedded feed to should faithfully reproduce the look & feel of the official Mastodon.  Doing so communicates clearly that it is Mastodon, and probably also gives viewers (false heheh) confidence that the feed is not fake, and also it is probably necessary for good visual appearance if we want to support advanced features like cards.  

### Is it possible to share anything between our implementations?

Thinking about this possibility, a library like `emfed` could perhaps be decomposed into several pieces:

1. A client API that can efficiently fetch the Mastodon feed from the network, with good error handling, and TypeScript typings for the returned objects
2. A renderer that faithfully recreates the official Mastodon look. It can have parameters (color theme, show "boost" button, etc) but making it extremely flexible seems not too useful, since in practice writing your own renderer is probably easier than restyling someone else's CSS.
3. For lightweight sites, a premade bundle that can be loaded using a simple `<script>` tag (as your current implementation). And maybe a free CDN URL for people who don't want to host the bundle themselves.

#3 is probably the least useful piece for my case.

But if #2 could be solved well using your minimal renderer, it could make sense for the React component to wrap that (rather than relying on React for rendering the entire feed). That could work.

Also, if you had a way to extract #1 into its own entry point or a separate NPM package, that could be useful also.

**Comment by sampsyo** [comment]
That's awesome! Glad it worked out, and it's also cool that you were able to recreate the "official" Mastodon appearance in your client.

It's a good idea to separate these various layers out. I'm going to abandon the approach prototyped in this PR and explore that separation as a way to make stuff more reusable—and hopefully easier to repurpose for different contexts in general.
